### PR TITLE
Eligible Author Origin-declared properties do not disable native appearance of elements

### PR DESCRIPTION
**Pull request by rr-codes**
#### e5dc8a60f80764c73b2a94c53a2f2ce891846906
<pre>
Eligible Author Origin-declared properties do not disable native appearance of elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=246980">https://bugs.webkit.org/show_bug.cgi?id=246980</a>
rdar://101521275

Reviewed by NOBODY (OOPS!).

Per the &quot;Properties Disabling Native Appearance&quot; spec (<a href="https://drafts.csswg.org/css-ui-4/#appearance-switching)">https://drafts.csswg.org/css-ui-4/#appearance-switching)</a>,
several properties should disable the native appearance of HTML elements when
declared in the author origin.

However currently, we do not do this, and instead only disable the native
appearance when the style doesn&apos;t match the style of the user agent. This PR
changes this behavior to match the specification.

* LayoutTests/TestExpectations:
* LayoutTests/fast/forms/auto-fill-button/input-auto-filled-and-obscured.html:
* LayoutTests/fast/forms/hidpi-textfield-background-bleeding-expected.html: Removed.
* LayoutTests/fast/forms/hidpi-textfield-background-bleeding.html: Removed.
* LayoutTests/fast/forms/input-background-ua-media-query.html:
* LayoutTests/fast/forms/select-style.html:
* LayoutTests/fast/selectors/text-field-selection-stroke-color-expected.html:
* LayoutTests/fast/selectors/text-field-selection-text-shadow-expected.html:
* LayoutTests/fast/selectors/text-field-selection-window-inactive-stroke-color-expected.html:
* LayoutTests/fast/selectors/text-field-selection-window-inactive-text-shadow-expected.html:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/fast/forms/select-style-expected.txt:
* LayoutTests/platform/win/TestExpectations:
* Source/WTF/wtf/cocoa/RuntimeApplicationChecksCocoa.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/makeprop.pl:
(generateColorValueSetter):
(generateInitialValueSetter):
(generateInheritValueSetter):
(generateValueSetter):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::isControlStyled const):
* Source/WebCore/rendering/RenderThemeCocoa.h:
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::isControlStyled const):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::isControlStyled const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::isControlStyled const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::hasAuthorOriginPropertyDisablingAppearance const):
(WebCore::RenderStyle::setHasAuthorOriginPropertyDisablingAppearance):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyInitialFontFeatureSettings):
(WebCore::Style::BuilderCustom::applyInheritFontFeatureSettings):
(WebCore::Style::BuilderCustom::applyInitialZoom):
(WebCore::Style::BuilderCustom::applyInheritZoom):
(WebCore::Style::BuilderCustom::applyInheritVerticalAlign):
(WebCore::Style::BuilderCustom::applyInheritSize):
(WebCore::Style::BuilderCustom::applyInitialSize):
(WebCore::Style::BuilderCustom::applyInheritTextIndent):
(WebCore::Style::BuilderCustom::applyInitialTextIndent):
(WebCore::Style::BuilderCustom::applyInheritLetterSpacing):
(WebCore::Style::BuilderCustom::applyInitialLetterSpacing):
(WebCore::Style::BuilderCustom::applyInheritLineHeight):
(WebCore::Style::BuilderCustom::applyInitialLineHeight):
(WebCore::Style::BuilderCustom::applyInheritListStyleType):
(WebCore::Style::BuilderCustom::applyInitialListStyleType):
(WebCore::Style::BuilderCustom::applyInheritOutlineStyle):
(WebCore::Style::BuilderCustom::applyInitialOutlineStyle):
(WebCore::Style::BuilderCustom::applyInitialCaretColor):
(WebCore::Style::BuilderCustom::applyInheritCaretColor):
(WebCore::Style::BuilderCustom::applyInitialClip):
(WebCore::Style::BuilderCustom::applyInheritClip):
(WebCore::Style::BuilderCustom::applyValueClip):
(WebCore::Style::BuilderCustom::applyInitialTextShadow):
(WebCore::Style::BuilderCustom::applyInheritTextShadow):
(WebCore::Style::BuilderCustom::applyInitialBoxShadow):
(WebCore::Style::BuilderCustom::applyInheritBoxShadow):
(WebCore::Style::BuilderCustom::applyInitialWebkitBoxShadow):
(WebCore::Style::BuilderCustom::applyInheritWebkitBoxShadow):
(WebCore::Style::BuilderCustom::applyInitialFontFamily):
(WebCore::Style::BuilderCustom::applyInheritFontFamily):
(WebCore::Style::BuilderCustom::applyInitialBorderBottomLeftRadius):
(WebCore::Style::BuilderCustom::applyInheritBorderBottomLeftRadius):
(WebCore::Style::BuilderCustom::applyValueBorderBottomLeftRadius):
(WebCore::Style::BuilderCustom::applyInitialBorderBottomRightRadius):
(WebCore::Style::BuilderCustom::applyInheritBorderBottomRightRadius):
(WebCore::Style::BuilderCustom::applyValueBorderBottomRightRadius):
(WebCore::Style::BuilderCustom::applyInitialBorderTopLeftRadius):
(WebCore::Style::BuilderCustom::applyInheritBorderTopLeftRadius):
(WebCore::Style::BuilderCustom::applyValueBorderTopLeftRadius):
(WebCore::Style::BuilderCustom::applyInitialBorderTopRightRadius):
(WebCore::Style::BuilderCustom::applyInheritBorderTopRightRadius):
(WebCore::Style::BuilderCustom::applyValueBorderTopRightRadius):
(WebCore::Style::BuilderCustom::applyInitialFontVariationSettings):
(WebCore::Style::BuilderCustom::applyInheritFontVariationSettings):
(WebCore::Style::BuilderCustom::applyInheritDisplay):
(WebCore::Style::BuilderCustom::applyInitialTextEmphasisStyle):
(WebCore::Style::BuilderCustom::applyInheritTextEmphasisStyle):
(WebCore::Style::BuilderCustom::applyInitialAspectRatio):
(WebCore::Style::BuilderCustom::applyInheritAspectRatio):
(WebCore::Style::BuilderCustom::applyInitialContain):
(WebCore::Style::BuilderCustom::applyInheritContain):
(WebCore::Style::BuilderCustom::applyInitialCounterIncrement):
(WebCore::Style::BuilderCustom::applyInheritCounterIncrement):
(WebCore::Style::BuilderCustom::applyInitialCounterReset):
(WebCore::Style::BuilderCustom::applyInheritCounterReset):
(WebCore::Style::BuilderCustom::applyInitialCursor):
(WebCore::Style::BuilderCustom::applyInheritCursor):
(WebCore::Style::BuilderCustom::applyInitialFill):
(WebCore::Style::BuilderCustom::applyInheritFill):
(WebCore::Style::BuilderCustom::applyInitialStroke):
(WebCore::Style::BuilderCustom::applyInheritStroke):
(WebCore::Style::BuilderCustom::applyInitialContent):
(WebCore::Style::BuilderCustom::applyInheritContent):
(WebCore::Style::BuilderCustom::applyInheritFontVariantLigatures):
(WebCore::Style::BuilderCustom::applyInitialFontVariantLigatures):
(WebCore::Style::BuilderCustom::applyInheritFontVariantNumeric):
(WebCore::Style::BuilderCustom::applyInitialFontVariantNumeric):
(WebCore::Style::BuilderCustom::applyInheritFontVariantEastAsian):
(WebCore::Style::BuilderCustom::applyInitialFontVariantEastAsian):
(WebCore::Style::BuilderCustom::applyInheritFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyInitialFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyValueFontVariantAlternates):
(WebCore::Style::BuilderCustom::applyInitialFontSize):
(WebCore::Style::BuilderCustom::applyInheritFontSize):
(WebCore::Style::BuilderCustom::applyInitialFontStyle):
(WebCore::Style::BuilderCustom::applyInheritFontStyle):
(WebCore::Style::BuilderCustom::applyInitialGridTemplateAreas):
(WebCore::Style::BuilderCustom::applyInheritGridTemplateAreas):
(WebCore::Style::BuilderCustom::applyValueGridTemplateAreas):
(WebCore::Style::BuilderCustom::applyInitialGridTemplateColumns):
(WebCore::Style::BuilderCustom::applyInheritGridTemplateColumns):
(WebCore::Style::BuilderCustom::applyInitialGridTemplateRows):
(WebCore::Style::BuilderCustom::applyInheritGridTemplateRows):
(WebCore::Style::BuilderCustom::applyInitialContainIntrinsicWidth):
(WebCore::Style::BuilderCustom::applyInheritContainIntrinsicWidth):
(WebCore::Style::BuilderCustom::applyInitialContainIntrinsicHeight):
(WebCore::Style::BuilderCustom::applyInheritContainIntrinsicHeight):
* Source/WebCore/style/StyleBuilderState.h:
(WebCore::Style::BuilderState::isAuthorOrigin const):
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:
(JSONCSSPropertiesChecker.check_codegen_properties):
</pre>
https://github.com/WebKit/WebKit/commit/e5dc8a60f80764c73b2a94c53a2f2ce891846906

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95193 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4345 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28441 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104790 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165049 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99184 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4446 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33476 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87511 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100686 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100859 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3239 "Found 30 new test failures: css2.1/20110323/replaced-elements-001.htm, css3/flexbox/flexitem.html, editing/selection/ios/select-text-under-hidden-subframe.html, fast/canvas/image-pattern-rotate.html, fast/css/text-input-with-webkit-border-radius.html, fast/forms/button-style-color.html, fast/forms/datalist/datalist-fallback-content.html, fast/forms/select-background-none.html, fast/forms/select-style.html, fast/scrolling/ios/click-events-during-momentum-scroll-in-main-frame.html ... (failure)") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81983 "Built successfully") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30207 "Found 34 new test failures: fast/forms/select-style.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-search-border-bottom-left-radius-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-search-border-bottom-right-radius-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-search-border-end-end-radius-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-search-border-end-start-radius-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-search-border-start-end-radius-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-search-border-start-start-radius-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-search-border-top-left-radius-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-search-border-top-right-radius-001.html, imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-generated/kind-of-widget-fallback-input-search-text-border-bottom-left-radius-001.html ... (failure)") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/98780 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85142 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87223 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73107 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/86295 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38922 "Built successfully") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18777 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-no-fallback-props-001.html (failure)") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/81559 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36739 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20072 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-no-fallback-props-001.html (failure)") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28067 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40676 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/42715 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-ui/compute-kind-widget-no-fallback-props-001.html, storage/indexeddb/crash-on-getdatabases.html (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/84233 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42661 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39313 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19034 "Passed tests") |

### CODE REVIEW

File: Source/WebCore/css/makeprop.pl [hunk 1]
```diff
@@ -1558,10 +1570,18 @@ sub generateValueSetter {
...
     }
     $setterContent .= $indent . "    " . generateSetValueStatement($name, $convertedValue) . ";\n";
+    $didUseValue = 1;
   }
   if (exists($propertiesWithStyleBuilderOptions{$name}{"fast-path-inherited"})) {
     $setterContent .= $indent . "    builderState.style().setDisallowsFastPathInheritance();\n";
   }
+  if (exists $propertiesWithStyleBuilderOptions{$name}{"is-appearance-disabling-property"}) {
+    if (!($didUseValue or $didCallSetValue)) {
+      $setterContent .= $indent . "    UNUSED_PARAM(value);\n";
```
Comment (pxlcoder) [inline Source/WebCore/css/makeprop.pl]: Why is this necessary for `is-appearance-disabling-property` and not the others?

File: Source/WebCore/rendering/RenderTheme.cpp [hunk 2]
```diff
@@ -834,8 +838,14 @@ bool RenderTheme::isControlStyled(const RenderStyle& style, const RenderStyle& u
     // FIXME: SearchFieldPart should be included here when making search fields style-able.
     case TextFieldPart:
     case TextAreaPart:
-        // Test the style to see if the UA border and background match.
-        return !style.borderAndBackgroundEqual(userAgentStyle);
+#if PLATFORM(COCOA)
```
Comment (pxlcoder) [inline Source/WebCore/rendering/RenderTheme.cpp]: Rather than add `PLATFORM(COCOA)` to a platform-agnostic file, we should only perform the link check in our platform-specific files (`RenderThemeMac` / `RenderThemeIOS`, or more ideally `RenderThemeCocoa`). 

You can move / adjust / copy the logic as needed.

File: Source/WebCore/rendering/RenderThemeIOS.mm [hunk 3]
```diff
@@ -461,17 +462,19 @@ static void drawJoinedLines(CGContextRef context, const Vector<CGPoint>& points,
 
 bool RenderThemeIOS::isControlStyled(const RenderStyle& style, const RenderStyle& userAgentStyle) const
 {
-    // Buttons and MenulistButtons are styled if they contain a background image.
-    if (style.effectiveAppearance() == PushButtonPart || style.effectiveAppearance() == MenulistButtonPart)
-        return !style.visitedDependentColor(CSSPropertyBackgroundColor).isVisible() || style.backgroundLayers().hasImage();
+    if (!inkedOnOrAfterSDKWithBehavior(SDKAlignedBehavior::AuthorOriginBorderBackgroundDeclarationsDisableAppearance)) {
```
Comment (pxlcoder) [inline Source/WebCore/rendering/RenderThemeIOS.mm]: `inkedOnOrAfterSDKWithBehavior`

File: Source/WebCore/rendering/style/RenderStyle.h [hunk 4]
```diff
@@ -380,6 +380,10 @@ class RenderStyle {
     std::optional<FontSelectionValue> fontItalic() const { return fontDescription().italic(); }
     FontPalette fontPalette() const { return fontDescription().fontPalette(); }
 
+    bool hasAuthorOriginPropertyDisablingAppearance() const { return m_rareNonInheritedData->hasAuthorOriginPropertyDisablingAppearance; } 
+
```
Comment (pxlcoder) [inline Source/WebCore/rendering/style/RenderStyle.h]: No need for a newline as these are related methods.

**Review by pxlcoder — COMMENTED** [review]
We will need to update TestExpectations.

File: Source/WebCore/css/makeprop.pl (hunk 1 above)
Comment (rr-codes) [reply Source/WebCore/css/makeprop.pl]: because all the others actually use `value`

Comment (pxlcoder) [reply Source/WebCore/css/makeprop.pl]: But this code is additive? I'm not sure I see the scenario where `value` is unused.

Can you share an example?
